comms/core/src/peer_manager/storage/database.rs

```diff
@@ -1192,6 +1165,7 @@ impl PeerDatabaseSql {
                     .or(peers::banned_until.lt(chrono::Utc::now().naive_utc())),
             )
             .filter(peers::deleted_at.is_null())
+            .filter(multi_addresses::last_failed_reason.is_null())
```

Comment:
this is bad as well. We cant just filter on null. If a peer failed, and we dont have anyother addresses, we need to go back end try ones which are null.

Reply:
Please add this in, this will be an issue with smaller networks like esme and igor

Reply:
Changes made in the latest commit:
- We first try to select non-failed (3 x count as per the selection algo), honouring the exclusion list passed in.
- If we cannot select the target amount of non-failed peers, we ignore the exclusion list and choose from the remaining peers at random.
After the initial selection, the algo will choose the best `count` number of peers from those selected for proactive dialling.

Reply:
we cant use random, it excludes the peers
`.filter(multi_addresses::last_seen.is_not_null())`
Just set a bypass for this filter when we need to exlcude it:
`.filter(multi_addresses::last_failed_reason.is_null())`

Reply:
Notes: 
- I implemented randomness and bypass failed in `PeerDatabaseSql::get_available_dial_candidates`.
- There was also a bug in `PeerDatabaseSql::get_available_dial_candidates`, whereby it did not return the selected number of peers.